core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java

```diff
@@ -376,20 +378,26 @@ public ClusterState execute(ClusterState currentState) {
                 while (iterator.hasNext()) {
                     Map.Entry<DiscoveryNode, List<MembershipAction.JoinCallback>> entry = iterator.next();
                     final DiscoveryNode node = entry.getKey();
-                    joinCallbacksToRespondTo.addAll(entry.getValue());
                     iterator.remove();
-                    if (currentState.nodes().nodeExists(node.getId())) {
+                    final String preflight = nodesBuilder.preflightPut(node);
+                    if (preflight != null) {
+                        Throwable failure = new IllegalStateException(preflight);
+                        for (MembershipAction.JoinCallback callback: entry.getValue()) {
+                            joinCallbacksToFail.add(new Tuple<>(callback, failure));
+                        }
+                    } else if (currentState.nodes().nodeExists(node.getId())) {
```

Comment:
One more question: What about `DiscoveryNodes.delta()`? It does not detect if node has been removed and readded. It also does not take `equalsIncludingMetaData` into account. This means that other code that relies upon the nodesDelta API in ClusterChangedEvent to detect if node has restarted might miss a restart? One such example is possibly `SnapshotsService.removedNodesCleanupNeeded`.